File: python/sdk/src/t0_provider_sdk/provider/validate_response.py
"""ConnectRPC interceptor that validates provider responses against buf.validate rules.

Invalid responses are rejected with Code.INTERNAL since they indicate
a provider implementation bug, not a client error.

The interceptor also emits one ``error``-level log line before re-raising so
providers see the failure in their own logs even when they don't wrap their
responses with :func:`t0_provider_sdk.provider.validate.validate`. The logger
is overridable via the constructor; the default is
``logging.getLogger("t0_provider_sdk")``.

Go equivalent: provider/validate_response.go
"""

from __future__ import annotations

import logging
from collections.abc import Awaitable, Callable
from typing import Any

import protovalidate
from connectrpc.code import Code
from connectrpc.errors import ConnectError
from connectrpc.request import RequestContext

from t0_provider_sdk._version import __version__
from t0_provider_sdk.provider.validate import _get_validator

DEFAULT_LOGGER_NAME = "t0_provider_sdk"


def _rpc_method_from_ctx(ctx: Any) -> str:
    """Best-effort extraction of the RPC method FQN from a RequestContext."""
    for attr in ("method", "procedure", "path"):
        value = getattr(ctx, attr, None)
        if isinstance(value, str) and value:
            return value
    return ""


def _log_validation_failure(
    logger: logging.Logger,
    response: Any,
    ctx: Any,
    error: protovalidate.ValidationError,
) -> None:
    """Emit one error-level line with structured fields for a validation failure."""
    response_type = (
        type(response).DESCRIPTOR.full_name if hasattr(type(response), "DESCRIPTOR") else type(response).__name__
    )
    logger.error(
        "response validation failed: %s",
        error,
        extra={
            "rpc_method": _rpc_method_from_ctx(ctx),
            "response_type": response_type,
            "violations": str(error),
            "sdk_version": __version__,
        },
    )


class ValidationInterceptor:
    """Async ConnectRPC unary interceptor that validates responses against proto rules."""

    def __init__(self, logger: logging.Logger | None = None) -> None:
        self._validator = _get_validator()
        self._logger = logger if logger is not None else logging.getLogger(DEFAULT_LOGGER_NAME)

    async def intercept_unary(
        self,
        call_next: Callable[[Any, RequestContext], Awaitable[Any]],
        request: Any,
        ctx: RequestContext,
    ) -> Any:
        response = await call_next(request, ctx)
        try:
            self._validator.validate(response)
        except protovalidate.ValidationError as e:
            _log_validation_failure(self._logger, response, ctx, e)
            raise ConnectError(Code.INTERNAL, f"response validation failed: {e}") from e
        return response


class ValidationInterceptorSync:
    """Sync ConnectRPC unary interceptor that validates responses against proto rules."""

    def __init__(self, logger: logging.Logger | None = None) -> None:
        self._validator = _get_validator()
        self._logger = logger if logger is not None else logging.getLogger(DEFAULT_LOGGER_NAME)

    def intercept_unary_sync(
        self,
        call_next: Callable[[Any, RequestContext], Any],
        request: Any,
        ctx: RequestContext,
    ) -> Any:
        response = call_next(request, ctx)
        try:
            self._validator.validate(response)
        except protovalidate.ValidationError as e:
            _log_validation_failure(self._logger, response, ctx, e)
            raise ConnectError(Code.INTERNAL, f"response validation failed: {e}") from e
        return response

File: go/provider/validate_response.go
package provider

import (
	"context"
	"log/slog"

	"buf.build/go/protovalidate"
	"connectrpc.com/connect"
	"connectrpc.com/validate"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protoreflect"

	"github.com/t-0-network/provider-sdk/go/sdkversion"
)

// newValidationInterceptor creates a ConnectRPC interceptor that validates
// provider responses against buf.validate proto annotations. Invalid responses
// are reported as errors with connect.CodeInternal — the wire shape is
// produced by connectrpc.com/validate and must remain unchanged.
//
// The returned interceptor adds one structured slog.Error line per validation
// failure (with rpc_method, response_type, violations, sdk_version fields) so
// providers see the failure even when their handler skipped the Validate
// helper. The original error is re-returned verbatim so callers downstream
// observe the same CodeInternal status and message.
func newValidationInterceptor(logger *slog.Logger) connect.Interceptor {
	inner := validate.NewInterceptor(validate.WithValidateResponses())
	return &loggingValidationInterceptor{
		inner:  inner,
		logger: logger,
	}
}

type loggingValidationInterceptor struct {
	inner  connect.Interceptor
	logger *slog.Logger
}

func (i *loggingValidationInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
	wrapped := i.inner.WrapUnary(next)
	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
		resp, err := wrapped(ctx, req)
		if err != nil {
			i.logIfResponseValidationFailure(req.Spec(), resp, err)
		}
		return resp, err
	}
}

func (i *loggingValidationInterceptor) WrapStreamingClient(next connect.StreamingClientFunc) connect.StreamingClientFunc {
	return i.inner.WrapStreamingClient(next)
}

func (i *loggingValidationInterceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
	return i.inner.WrapStreamingHandler(next)
}

// logIfResponseValidationFailure emits a single error-level line when err is
// the response-validation failure produced by connectrpc.com/validate. Request
// validation failures (CodeInvalidArgument) and any other handler errors are
// ignored so we don't spam the log on unrelated failure paths.
func (i *loggingValidationInterceptor) logIfResponseValidationFailure(
	spec connect.Spec,
	resp connect.AnyResponse,
	err error,
) {
	if connect.CodeOf(err) != connect.CodeInternal {
		return
	}
	ve := asValidationError(err)
	if ve == nil {
		return
	}

	logger := i.logger
	if logger == nil {
		logger = slog.Default()
	}
	logger.Error(
		"response validation failed",
		slog.String("rpc_method", spec.Procedure),
		slog.String("response_type", responseTypeName(spec, resp)),
		slog.Any("violations", formatViolations(ve)),
		slog.String("sdk_version", sdkversion.Version),
	)
}

// responseTypeName returns the proto FQN of the response message. It prefers
// the actual response payload when present (success / partial-response paths),
// and falls back to the procedure's MethodDescriptor when the inner validate
// interceptor returned a nil response on failure. Returns "unknown" only when
// neither source carries proto metadata — which would indicate a non-protobuf
// custom transport.
func responseTypeName(spec connect.Spec, resp connect.AnyResponse) string {
	if resp != nil {
		if msg, ok := resp.Any().(proto.Message); ok {
			return string(msg.ProtoReflect().Descriptor().FullName())
		}
	}
	if md, ok := spec.Schema.(protoreflect.MethodDescriptor); ok {
		return string(md.Output().FullName())
	}
	return "unknown"
}

// formatViolations renders the violations list as a slice of short strings so
// slog handlers can serialise it as either a JSON array or a textual list
// without holding references to protobuf reflection objects.
func formatViolations(ve *protovalidate.ValidationError) []string {
	if ve == nil {
		return nil
	}
	out := make([]string, 0, len(ve.Violations))
	for _, v := range ve.Violations {
		out = append(out, v.String())
	}
	return out
}
